### ISSUE TITLE
Copy of E2E test support for HW wallet support -> discovery

### ISSUE THREAD

**Issue by angelcheung22** [MetaMask/metamask-mobile#9544]
Add validation for the Ledger and HW wallet support in mobile (and extension) within the E2E tests?
In the past we've discussed the potential of using software emulators (existing or custom) to mock the responses from the device as needed at the URL or library level. 

We can try mock @ledgerhq/react-native-hw-transport-ble to bypass the bluetooth communication. and then we can try to see how to do it in e2e tests. the only downside is if library upgrade, we may need to change the mock again to refect the library API change.

**Comment by angelcheung22** [comment]
Pretty stuck as the mocking failed. and MM QA team has a meeting on 20 June to discuss overall QA strategy, so put on hold

**Comment by Akaryatrh** [comment]
**Quick statement on our findings:**
- e2e tests are using Detox on mobile side
- Detox doesn't have any mocking feature included, so mocking should be done on Metro's side (more info here https://wix.github.io/Detox/docs/guide/mocking)

But when doing so, we're facing an error with module loading.

I have setup a branch to test mocking strategy here:  https://github.com/MetaMask/metamask-mobile/tree/feat/e2e-ledger

Instructions:
- You need to run first: `yarn && yarn setup`
- then, open a new tab and run `yarn watch:e2e`
- Then come back to first tab and run `yarn test:e2e:android:debug:build`
- Once done, you can run `yarn test:e2e:android:debug:run e2e/specs/hardware-wallets/add-ledger-device.spec.js`

If you have the same issue as me, you'll see that app is crashing :confused:
<img width="389" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBeVNpQlE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--2fa53a7392cfb4aaed693811688c51fecdfad235/Capture%20d%E2%80%99e%CC%81cran%202024-06-10%20a%CC%80%2014.55.44%20%281%29.png" alt="Capture d’écran 2024-06-10 à 14.55.44 (1).png" />


**How mocks are working:**
Basically It's needed to duplicate a file and add the `mock.ts`  extension instead of `.ts`  beside the original file. Then metro builder should take the mock file instead. Related logic based on env variable `IS_TEST` can be seen inside `metro.config.js`
Some hooks have been mocked just to test if mocking strategy was working. But at the end we would like to mock the bluetooth library only (potentially `react-native-ble-plx` or lower)

**Comment by Akaryatrh** [comment]
Manual tests can be found here https://github.com/MetaMask/metamask-extension/tree/develop/test/manual-scenarios and those specific for hardware wallets are here: https://github.com/MetaMask/metamask-extension/tree/develop/test/manual-scenarios/keyring

As there are not many specific test for HW, i guess we can reuse the ones for regular accounts (or at least get come inspiration)

**Comment by Akaryatrh** [comment]
Please find here testing scenarios for Ledger in the context of MM ext https://www.notion.so/consensys/66dff75372c2438b833a7c6575afa13a?v=86c94a328b0d4bb7981e9b51f34161a3

**Comment by Montoya** [comment]
If we could do anything to catch this type of error in an E2E test, it may be to assert that the payload used in a hardware wallet approval has not changed. 

1. Find a payload we can use in testing, like "Swap USDC for DAI on Ethereum Mainnet"
2. Create an E2E test where we assert that the payload has not changed from our example to the RC
3. Throw if the payload has changed in any way

Can write a second test comparing QR codes as well, so 1 test for raw payload and another for QR code generated

**Comment by Akaryatrh** [comment]
**Update on upcoming changes:**
- new [PR](https://github.com/MetaMask/metamask-extension/pull/33048) opened by @seaona to cover basic tests for ledger (also migrates to page objects pattern)
- once the above merged, we can resume the work started here https://github.com/MetaMask/metamask-extension/pull/27966 (tx related tests) and here https://github.com/MetaMask/metamask-extension/pull/27930 (signature related tests)
- once done, we can try to add back swap related tests
